qiskit/extensions/unitary.py

```diff
@@ -111,6 +111,48 @@ def _define(self):
             raise NotImplementedError("Not able to generate a subcircuit for "
                                       "a {}-qubit unitary".format(self.num_qubits))
 
+    def control(self, num_ctrl_qubits=1, label=None):
+        """Return controlled version of gate
+
+        Args:
+            num_ctrl_qubits (int): number of controls to add to gate (default=1)
+            label (str): optional gate label
+
+        Returns:
+            UnitaryGate: controlled version of gate.
+
+        Raises:
+            QiskitError: unrecognized mode
+        """
+        cmat = self._compute_control_matrix(self.to_matrix(), num_ctrl_qubits)
+        return UnitaryGate(cmat, label=label)
+
+    def _compute_control_matrix(self, base_mat, num_ctrl_qubits, phase=0):
+        """
+        Compute the controlled version of the input matrix with qiskit ordering.
+
+        Args:
+            base_mat (ndarray): unitary to be controlled
+            num_ctrl_qubits (int): number of controls for new unitary
+            phase (float): The global phase of base_mat which is promoted to the
+                global phase of the controlled matrix
+
+        Returns:
+            ndarray: controlled version of base matrix.
+        """
+        num_target = int(numpy.log2(base_mat.shape[0]))
+        ctrl_dim = 2**num_ctrl_qubits
+        ctrl_grnd = numpy.repeat([[1], [0]], [1, ctrl_dim-1])
+        full_mat_dim = ctrl_dim * base_mat.shape[0]
+        full_mat = numpy.zeros((full_mat_dim, full_mat_dim), dtype=base_mat.dtype)
+        for i in range(ctrl_dim-1):
+            full_mat += numpy.kron(numpy.eye(2**num_target),
+                                   numpy.diag(numpy.roll(ctrl_grnd, i)))
+        if phase != 0:
+            full_mat = numpy.exp(1j * phase) * full_mat
```

Comment:
This looks wrong. If there is a a phase on the unitary being controlled, the final matrix should be: ` id \otimes (id - |1,...,1><1,...,1|)  + (exp(i phase)*U) \otimes |1,...,1><1,...,1|`.

I would just remove option of phase in this PR, you don't really need it for unitary matrix because you can just include it in the matrix itself (and i can make sure it's handled properly in #3472).

Also rather than the for loop you can just construct this matrix as two blocks using the above example:

```python
ctrl_proj = numpy.diag(numpy.roll(ctrl_grnd, ctrl_dim - 1)))
full_mat = numpy.kron(numpy.eye(2**num_target), numpy.eye(ctrl_dim) - ctrl_proj) + numpy.kron(base_mat, ctrl_proj)
```

Reply:
I'll remove it for this pr.